Implement snapshot replication.

FransUrbo opened the pull request:
Not actually tested, other than in dry-run mode. But the commands it would run seems to be correct.

Sign-off: Turbo Fredriksson turbo@bayour.com
Closes: #5

FransUrbo committed:
Implement snapshot replication.

wankdanker commented:
I would just like to point out @mk01's [fork of zfs-auto-snapshot](https://github.com/mk01/zfs-auto-snapshot). It has an implementation of snapshot replication with some nice options:

```text
    -i, --send-at-once Send more incremental snapshots at once in one package
                       (-i argument is passed to zfs send instead of -I). 
    --send-full=F      Send zfs full backup. F is target filesystem.
    --send-incr=F      Send zfs incremental backup. F is target filesystem.
    -X, --destroy      Destroy remote snapshots to allow --send-full if 
                       destination has snapshots (needed for -F in case 
                       incremental snapshots on local and remote do not match).
                       -f is used automatically.
    -F, --fallback     Allow fallback from --send-incr to --send-full, 
                       if incremental sending is not possible (filesystem 
                       on remote just created or snapshots do not match - 
                       see -X).  
    -R, --replication  Use zfs's replication (zfs send -R) instead of simple 
                       send over newly created snapshots (check man zfs for 
                       details). -f is used automatically. 
```
Maybe it's not the right fit for the official distribution of zfs-auto-snapshot but thought it is worth mentioning.

FransUrbo commented:
Diffing the two, @mk01 basically rewrote the whole thing.. Which makes it impossible to merge.

FransUrbo committed:
Implement --{pre,post}-send option for the snapshot replication.

wankdanker commented:
OK. Thank you for taking a look.

kpande commented:
I based my version of zfs-auto-replicate on @mk01's script as well. It has benefits like being cross-platform with FreeBSD and Darwin (Mac) machines.

FransUrbo committed:
Implement --send-{ssh,mbuf}-opts for the snapshot replication.

FransUrbo commented:
Having options to <code>ssh</code> was a nice touch though. Added that. And the possibility to use <code>mbuffer</code>.

I still haven't decided if the fallback is a good idea or not. It's usually only used once (the first time) and this tool seems more towards incrementals (a snapshot IS after all an incremental change - from the previous one). It's not a difficult change, but I just don't know if it's a good idea. 

Could @wankdanker and/or @kpande give me some examples where this is used (often)?

kpande commented:
My use of the script used it often because the receiving side is used as an active node and then it will destroy all volumes and receive them again.. otherwise it complains.

FransUrbo commented:
Then why don't you use <code>--send-full</code> instead. Because that is essentially the result… ?

kpande commented:
because i wanted incremental to happen when at all possible. sometimes it wasn't, and you have to destroy everything to redo it. in my case it was fairly often that i had to do this - not frequently enough to only use send-full but often enough to never do it manually.

FransUrbo commented:
Ok, fair enough. It's an easy change anyway.

wankdanker commented:
In sending snapshots offsite it's best for us to only send the incremental changes over the internet. But if that fails, we still need to get the latest snapshot sent offsite, so destroy whatever is on the remote side and send full needs to be done. That's our use case.

kpande commented:
more desperately than anything this script needs a way to track send/recv successes and failures, and handle sending to multiple hosts. that is why i needed to use the fallback anyways, because it is not written to handle multiple hosts.. 

with multiple servers, you need to do multiple runs of z-a-s for each receiving host and then it creates a new snapshot on each execution and if you're using --keep X, with more than three receiving servers you will destroy snapshots that need to exist for host A and B to do incrementals properly.

z-a-s could have the option to use a simple cache db of sorts. the db can contain records of which hosts were attempted and failed so that it then has the ability to retry sends. 

if you don't want a db, then the "solution" for multiple server support is to have --keep X about equal or greater than the number of snapshots you'll be creating (i.e. keep 8 would cover maybe 7 servers? I haven't tested it that way)

even better for z-a-s would be an option to use whatever snapshots are already on the system. in essence z-a-s should never create or destroy snapshots. it would only compare a list of the sending and receiving system's existing snapshots and transfer the differences. this would useful if another tool is creating snapshots and managing their deletion and I just want to use z-a-s to manage replication of server states.
^ this is what i do using a PHP script for Jentu. I can't share it here (I wish I could!) and I can't write the equivalent for ZoL, but it's a great idea for z-a-s to be less intrusive.

FransUrbo committed:
Implement --send-fallback for the snapshot replication.
This allows a --send-incr to fallback to --send-full if it can't find
an earlier snapshot.

wankdanker commented:
I agree with @kpande on the option to not create snapshots and just send at least the most recent ones. It would be nice to have a --send-only or --skip-snapshots option that would bypass creating new snapshots and just send the most recently created ones.

FransUrbo commented:
As I just said to @kpande on IRC, I'm not sure that zfs-auto-snapshot is the tool for that... It's called '...snapshot' after all. Having simple replication is good, but I don't think it's in the tools ... 'mission statement' to do such advanced replication..

It handles snapshots. That's it's name, it's purpose. Having ANY replication _at all_ might be to much for the tool in the first place. Which is why I've tried to keep it simple. For those that want/need such advanced replication as you request, you should use a tool specifically geared towards replication.

wankdanker commented:
I hear you. And I agree with you than any replication at all should technically be out of the scope of the project based on its name. It's just so tempting to want it all in one tool. :)

FransUrbo commented:
Note that ALL my commits in this PR is based on what's on the local system. It _assumes_ that the receiving end is identical to localhost. if the remote end deletes and creates snapshots on it's own, my commits will fail anyway.

FransUrbo commented:
> It's just so tempting to want it all in one tool. :)
> 
How very un-UNIX of you. Shame! :)

kpande commented:
the receiving end isn't creating/deleting snapshots, they're static. the issue is that the sending part is creating a new snapshot for each receiving host. if the name is the issue, change it to zfs-auto-backup and have it do everything it needs to, because it's just a stones' throw from being really useful.

wankdanker commented:
> How very un-UNIX of you. Shame! :)

Yeah, that's the lazy sysadmin in me. Sounds like its time for zfs-auto-replicate. :)

FransUrbo commented:
> if the name is the issue, change it to zfs-auto-backup and have it do everything it needs to
> 
That is @dajhorn choice. I'm just not directly convinced _I_ want that to happen. I like the simplicity of the tool. It does what it says - manages snapshots. As in, rotates them so I don't have to many of them.

The fact that there is (limited) support for replication is enough (for me).

But if @dajhorn chooses to rename it and make it into a fully fledged backup and replication tool for ZFS, I might help out where needed. But as I've said, I don't really want that to happen. One tool, one job/purpose. That's the UNIX way :)

kpande commented:
The UNIX way is to do something and do it _well_, you're ignoring the last part by leaving out support for multiple recv hosts. It's even a simple addition that would fix it. You're hung up on the skipping snapshot part, but that's not the only purpose of the addition.. you can run z-a-s without skip-snapshot and then run it as many subsequent times as needed for the other receiving hosts and it'll send what it already just created. Makes Sense.

otherwise z-a-s should just be left as is. no need for half-baked features..

kpande commented:
for the record, @dajhorn last activity was 5 months ago, beginning to wonder if he is sick or unwell somewhere.

FransUrbo commented:
Having support for multiple receiving host is, as you say, easy. BUT, and it's a big but. It still *require* that each receiver is identical to the sender...

FransUrbo commented:
If you want support for differences between sender and receiver, it's going to be a _huge_ change.

I might consider it if (when! :) @dajhorn have merged these ones. But as you say, he's been 'lost' somewhere...

kpande commented:
that's what the tool is supposed to do, create an exact replica within the given recv path.. that is what I expect all ZFS snapshot replication tools to do..

FransUrbo commented:
Yet, your receiver(s) isn't. Because you use a hacked script? I really don't understand why your receiver(s) is so different, but if a snapshot is created on sender/localhost it, and *only* it, should be sent to the receiver(s)...

FransUrbo commented:
_I THINK_ that if the receiver(s) have come out of sync with the sender/localhost/master (whatever one should call it), it should be up to the administrator to manually fix this before using an *auto* tool.

kpande commented:
i don't understand what you're saying. you must be misunderstanding the issue.

I have three servers, in my house. If I use zfs-auto-snapshot to send / recv snapshots in between server A and B, it's fine. but when I run the tool again a minute later (or whenever it finishes replicating to the first host), it creates a new snapshot and sends the full dataset to server C. if I have --keep 2 because I want to keep the # of snapshots from z-a-s to a minimum, it will destroy the older one that was used to send to server B.

an hour later, let's say I do this again. Instead of sending incrementally to server B, it does a full send because there's missing snapshots on server A and the latest doesn't match. Great. Then do server C and it does the _same thing_. the script is absolutely useless because it keeps re-sending the full differences.

FransUrbo commented:
Ah, ok. I think I get it now. Because you have two receivers, but zfs-auto-snapshot can only send to one at a time, zfs-auto-snapshot **creates** the inconsistency…

Ok, fair enough. I'll add support for sending to multiple hosts. As I said, that part is fairly easy. Give me a couple of minutes.

kpande commented:
no problemo, just trying to be understood in the world is so difficult sometimes! must be a language barrier somewhere :-) :+1:

kpande commented:
Like I said on IRC:
do you suppose sending to multiple hosts at once is the best solution, because skip-snapshots would allow you to re-run a failed send too

FransUrbo commented:
What options would <code>--skip-snapshots</code> take? How would it be used?

kpande commented:
like I said above:

it would only compare a list of the sending and receiving system's existing snapshots and transfer the differences.

by differences, I mean, whichever new snapshots that have been created on the send side that don't exist yet on the recv side.

the whole --keep thing becomes difficult when dealing with multiple recv hosts..

wankdanker commented:
I can see the complication that adding `--skip-snapshots` would make to the actual code. Right now it looks like the list of snapshots to send is built as the snapshots themselves are created. 

The way I see it, when `--skip-snapshots` is set (no options that I can think of currently) it would behave similarly to how `--destroy-only` works. But instead of doing destroy logic for each target, it would attempt to send the most recent snapshot that exists on localhost. 

I think it might be more clear to call this option `--send-only`.

FransUrbo commented:
At the moment, I can't see <code>--skip-snapshots</code> being a reality. It would *require* a remote list of snapshots and filesystem and then go through and compare them.

That *in addition* the check we're already doing by comparing the *local* list of snapshots.

wankdanker commented:
Example usage might be:

```bash
$ zfs-auto-snapshot --send-incr=remote-server:remote-pool --send-only --send-fallback //
```

Here is what would happen:
* List of target file systems/volumes is generated
* For each target:
  * Determine most recent snapshot
  * Append most recent snapshot to list of snapshots to send
* Execute do_send() which abides by the --send incr and --send-fallback options

So, no new snapshots are created, only the most recent snapshots for each target are sent to remote-server.

wankdanker commented:
I just skimmed the code, so I am probably missing some of the additional complications. But, it doesn't seem like it is necessary to get a list of remote snapshots and file systems. For each most recent snapshot on localhost, attempt to send it to the remote. 

I'll dig deeper into your code and see if I can implement my vision.

FransUrbo commented:
Yes, but it should only do that if the sender and receiver is not in sync.

This is what it do now:

* Get a full list of all local filesystems and snapshots
* Create snapshots
  * Possibly remove old one
* Remember the snapshot being created
* For each of these snaps, find the last (local) one [before this new one]
  * If recursive is called, find all snaps with this date
* Send those snapshots

This is quite simple, and should work. It's not bullet proof, but if there's a failure, I still think it needs manual administrator fixing.

IF one would implement <code>--skip-snaps</code>, this would have to be done:

* Get a full list of all local filesystems and snapshots
* Create snapshots
  * Possibly remove [local] old one
* Remember the [local] snapshot being created
* For each of these [local] snaps, find the last [local] one [before this new one]
  * If recursive is called, find all [local] snaps with this date
* Send those snapshots
  * If this fails:
    * Get a full list of all remote filesystems and snapshots
    * Go through all remote snapshots, finding the last one
      * If recursive is called, find all [remote] snaps with this date
    * Send that snapshot again, with a different incremental base (gathered earlier)

and it might even be more complicated than that. Need to think about this some more.

kpande commented:
so your script doesn't clean up remote snapshots?

FransUrbo commented:
No it doesn't.

FransUrbo committed:
Implement multiple receive hosts for the snapshot replication.

FransUrbo commented:
So in practice, the only think not implemented from @mk01's version is the <code>--destroy</code> to destroy/cleanup remote snapshots…

Anything else?

wankdanker committed:
Implement --send-only for the snapshot replication.
Used for skipping snapshotting and only do the replication part.

wankdanker commented:
See https://github.com/wankdanker/zfs-auto-snapshot/commit/7ad310063ce00afe9cdcfc7ac21db2a80fc5bb10 for what I was picturing for --send-only. 

It generated the send command as I expected for something like:

```bash
$ ./zfs-auto-snapshot.sh -v -n --send-incr=bu01:test --send-only --send-fallback --label=hourly ms03-ssd/vm01-disk
```

FransUrbo commented:
@wankdanker Thanx! Looks good. I've merged that into this PR if you don't mind. If you do, let me know and I'll remove it (it's still marked as yours).

wankdanker commented:
@FransUrbo that's great, thank you. 

I didn't get the impression that @mk01's `--destroy` option actually cleaned up anything. I thought it would just destroy the remote target if it existed but couldn't receive because there were existing snapshots that didn't work out with the send stream.

It looks like you handle that with the `$remote -F $opt_recv_pool` in the case where fallback is allowed.

The one nice addition using the `--send-fallback` option would be if the incremental send fails because the remote filesystems don't have the correct snapshots then it should try again with a full send.

FransUrbo commented:
@wankdanker I broke out the part that gets the last snapshot into it's own function (see https://github.com/FransUrbo/zfs-auto-snapshot/commit/3b0e38385f16756f8e6e0e13cb59e068c25110aa#diff-e7f2480c37404caadf77e766166e1b3aR155) and then your addition was simple (see https://github.com/FransUrbo/zfs-auto-snapshot/commit/705ec33aae0382edb129d61bfe34323d6bff1d29#diff-e7f2480c37404caadf77e766166e1b3aR270).

FransUrbo committed:
Implement do_destroy_remotes() to destroy the exact same snapshots
on remote host(s) which is destroyed on localhost.

wankdanker commented:
Awesome work, @FransUrbo. :+1: :beers:

colttt commented:
how can it be used to send backups via cron automatically?
